File: app/installer.py
"""Install agents to AI tools and backup agents from tools.

Windows-only, experimental.
"""
from __future__ import annotations

from pathlib import Path
from typing import Callable

from .categories import ensure_category
from .storage import AgentSkill, save_skill
from .tool_registry import TOOL_BY_ID


def install_agents(
    skills: list[AgentSkill],
    tool_id: str,
    *,
    target: Path | None = None,
    progress_callback: Callable[[int, int, str], None] | None = None,
) -> tuple[int, int]:
    """Install skills to a tool. Returns (success_count, fail_count)."""
    tool = TOOL_BY_ID.get(tool_id)
    if not tool:
        return 0, len(skills)

    dest = target if target is not None else tool.default_path
    success = fail = 0

    for i, skill in enumerate(skills):
        try:
            tool.install(skill, dest)
            success += 1
        except Exception:
            fail += 1
        if progress_callback:
            progress_callback(i + 1, len(skills), skill.slug)

    return success, fail


def backup_from_tool(
    tool_id: str,
    *,
    source: Path | None = None,
    skip_existing: bool = True,
    progress_callback: Callable[[int, int, str], None] | None = None,
) -> tuple[int, int]:
    """Backup skills from a tool into agent-manager. Returns (imported, skipped)."""
    tool = TOOL_BY_ID.get(tool_id)
    if not tool:
        return 0, 0

    src = source if source is not None else tool.default_path
    skills = list(tool.backup_skills(src))
    imported = skipped = 0

    for i, skill in enumerate(skills):
        try:
            ensure_category(skill.path.parent.parent.name)
            if skip_existing and skill.path.exists():
                skipped += 1
            else:
                save_skill(skill, backup=False)
                imported += 1
        except Exception:
            skipped += 1
        if progress_callback:
            progress_callback(i + 1, len(skills), skill.slug)

    return imported, skipped

File: src/installer.rs
use std::path::Path;

use crate::{
    AppPaths,
    categories::ensure_category,
    model::AgentSkill,
    storage,
    tool_registry::{ToolKind, backup_skills, find_tool, install_skill, install_skills},
};

pub fn install_agents(skills: &[AgentSkill], tool_id: &str, target: &Path) -> (usize, usize) {
    let Some(tool) = find_tool(tool_id) else {
        return (0, skills.len());
    };
    if matches!(tool.kind, ToolKind::Aider | ToolKind::Windsurf) {
        return match install_skills(&tool, skills, target) {
            Ok(_) => (skills.len(), 0),
            Err(_) => (0, skills.len()),
        };
    }
    skills.iter().fold((0, 0), |(success, fail), skill| {
        match install_skill(&tool, skill, target) {
            Ok(_) => (success + 1, fail),
            Err(_) => (success, fail + 1),
        }
    })
}

pub fn backup_from_tool(
    tool_id: &str,
    source: &Path,
    paths: &AppPaths,
    skip_existing: bool,
) -> (usize, usize) {
    let Some(tool) = find_tool(tool_id) else {
        return (0, 0);
    };
    let mut imported = 0;
    let mut skipped = 0;
    for skill in backup_skills(&tool, source, paths) {
        if skip_existing && skill.path.exists() {
            skipped += 1;
            continue;
        }
        if let Some(category) = skill
            .path
            .parent()
            .and_then(|path| path.parent())
            .and_then(|path| path.file_name())
            .map(|name| name.to_string_lossy().into_owned())
        {
            let _ = ensure_category(&paths.agents, &category);
        }
        if storage::save_skill(&skill, paths, skill.path.exists()).is_ok() {
            imported += 1;
        } else {
            skipped += 1;
        }
    }
    (imported, skipped)
}
